feat(language-service): [Ivy] getSemanticDiagnostics for external templates

kyliau opened the pull request:
This PR enables `getSemanticDiagnostics()` to be called on external templates.

Several changes are needed to land this feature:

1. The adapter needs to implement two additional methods:
   a. `readResource()`
      Load the template from snapshot instead of reading from disk
   b. `getModifiedResourceFiles()`
      Inform the compiler that external templates have changed so that the
      loader could invalidate its internal cache.
2. Create `ScriptInfo` for external templates in MockHost.
   Prior to this, MockHost only track changes in TypeScript files. Now it
   needs to create `ScriptInfo` for external templates as well.

For (1), in order to make sure we don't reload the template if it hasn't
changed, we need to keep track of its version. Since the complexity has
increased, the adapter is refactored into its own class.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

kyliau committed:
feat(language-service): [Ivy] getSemanticDiagnostics for external templates

This PR enables `getSemanticDiagnostics()` to be called on external templates.

Several changes are needed to land this feature:

1. The adapter needs to implement two additional methods:
   a. `readResource()`
       Load the template from snapshot instead of reading from disk
   b. `getModifiedResourceFiles()`
       Inform the compiler that external templates have changed so that the
       loader could invalidate its internal cache.
2. Create `ScriptInfo` for external templates in MockHost.
   Prior to this, MockHost only track changes in TypeScript files. Now it
   needs to create `ScriptInfo` for external templates as well.

For (1), in order to make sure we don't reload the template if it hasn't
changed, we need to keep track of its version. Since the complexity has
increased, the adapter is refactored into its own class.

atscott commented on packages/language-service/ivy/language_service.ts:
```diff
@@ -21,31 +21,50 @@ export class LanguageService {
 
-  private createCompiler(program: ts.Program): NgCompiler {
+  private createCompiler(program: ts.Program, fileName: string): NgCompiler {
+    if (!isTypeScriptFile(fileName)) {
```
I feel like the way this is done is a bit awkward. 
* `this.modifiedResourceFiles` can never have more than one value, so a `Set` is kind of weird. 
* I feel a little weird about how the items in it are being used. The `NgCompilerAdapter` relies heavily on the reference to this not changing in order to function correctly. Sort of a combination of relying on the immutability of the reference and the mutability of the data.
* The value is only updated in one place but it ends up being escalated to a class level property just so the reference can be used in the adapter.

What if you did something like this:
```
  constructor(project: ts.server.Project, private readonly tsLS: ts.LanguageService) {
    this.adapter = createNgCompilerAdapter(project);
  }

  function createNgCompilerAdapter(project: ts.server.Project): NgCompilerAdapter {
    return {
      ...
      getModifiedResourceFiles(): Set<string> {
        return new Set();
      }
    }
  }

  private createCompiler(program: ts.Program, fileName: string): NgCompiler {
    if (!isTypeScriptFile(fileName)) {
      const modifiedResources = new Set([fileName]);
      this.adapter = {...this.adapter, getModifiedResourceFiles: () => modifiedResources};
    }
    ...
  }

```

kyliau replied:
Thank you for the excellent feedback!
While addressing your comment I realize there's a fatal flaw in this design: the external template is unconditionally invalidated on every single request, even if nothing has changed. I'll think of a better solution, and also spend some time to make sure we have tests to cover this scenario.

ayazhafiz commented on packages/language-service/ivy/language_service.ts:
```diff
@@ -21,31 +21,50 @@ export class LanguageService {
   private lastKnownProgram: ts.Program|null = null;
   private readonly strategy: TypeCheckingProgramStrategy;
   private readonly adapter: NgCompilerAdapter;
+  /**
+   * This is a persistent data structure to signal to the Ivy compiler that
+   * some of the external templates have changed, so that the resource loader
+   * would invalidate its cache and fetch the latest content from snapshot.
+   */
+  private readonly modifiedResourceFiles = new Set<string>();
```
I think it would be helpful to explain why keeping track of this is needed by the Ivy compiler. My assumption would be that the TS compiler does not inform us of changes to Angular templates so this is why we have to do it, but either way it could be helpful to clarify.

Also, is there a reason we want to leak the Resource abstraction here? My understanding is that the only modified resources that language service would manage as a ResourceHost would be HTML templates. If we are not managing anything else, I think it could be more clear to describe this member as modified template files. I realize this is pedantic though :)

kyliau replied:
I've moved the logic to a standalone "LanguageServiceAdapter" to make it easier to understand the role of "modifiedResourceFiles".
Also renamed "resource" to "template" wherever applicable since that's the only thing we care about in language service.

kyliau commented on packages/language-service/ivy/language_service.ts:
```diff
@@ -113,6 +140,24 @@ function createNgCompilerAdapter(project: ts.server.Project): NgCompilerAdapter
       return project.projectService.toCanonicalFileName(fileName);
     },
+    // readResource() and getModifiedResourceFiles() are methods on ExtendedTsCompilerHost
+    readResource(fileName: string): string {
```
done

kyliau commented on packages/language-service/ivy/language_service.ts:
```diff
@@ -21,31 +21,50 @@ export class LanguageService {
+      // If this is a resource file (external template), then add it to the
+      // modifiedResourceFiles set to signal the loader to invalidate its
+      // cache and reload the content from snapshot.
+      this.modifiedResourceFiles.clear();
```
It's cleared so that we don't do more work than necessary - although it's possible that more than one template has changed, most language service APIs only work with a single file at a time, so we want to make sure we don't propagate the change from the last request to the next.

kyliau commented:
@atscott @ayazhafiz As mentioned in the meeting just now, I've refactored the adapter into a standalone class, and added more comments to explain why we need to manage resource (templates) separately. PTAL.

ayazhafiz commented on packages/language-service/ivy/language_service_adapter.ts:
```diff
@@ -0,0 +1,112 @@
+   * It is a method on ExtendedTsCompilerHost, see
+   * packages/compiler-cli/src/ngtsc/core/api/src/interfaces.ts
+   */
+  getModifiedResourceFiles(): Set<string> {
+    return this.modifiedTemplates;
+  }
```
For example, could we clear the modifiedTemplates here and return the previous contents? (If the compiler only needs the same changes once)

kyliau replied:
I think it's not a good idea to clear the modified templates here, because
1. we are relying on the compiler to only call this method once
    This assumption is fragile. Within the compiler it's perfectly fine for it to call this method multiple times, and it'd be surprising to them if the results of the first call is different from the second call
2. the behavior would not be consistent within the _same_ compiler run
    This is a method defined by the compiler, and we need to adhere to the behavior they expect, not the behavior we provide.
https://github.com/angular/angular/blob/bc69182bdd2f084466a2f722afc85cde425e1454/packages/compiler-cli/src/ngtsc/core/api/src/interfaces.ts#L47-L51
I gave this some more thought, and realize that we could clear the modified template *after* the compiler has read the updated version. The pro is that it'll always be consistent, i.e. once the compiler reads the template, it's no longer marked as modified. The con is that there's no explicit way for language service to make sure the template update is cleared after every request. I think pro > con in this case, because it's very unlikely that the compiler would not read a template that's been marked as modified.
With this, I deleted the `clearTemplateUpdates()` method.

ayazhafiz replied:
One more note: let's say the compiler asks for all the modified resources (getModifiedResourceFiles), then reads one modified resource, then asks for the modified resources again. In this case, with what you have described, the second set of modified resources would be one less than the first, so the expectation that first call = second call in the same run is lost. Is this something we need to be concerned about?

kyliau replied:
I think the behavior is still consistent in this case, because once the compiler reads a template, it already has the latest state of the file, so the file is no longer considered "modified". 
That said, I agree this is not the best design. I think the initial design with `clearTemplateUpdates()` is the most straightforward solution with no surprise. Would you prefer that over the current solution?

ayazhafiz replied:
Why don't we just recreate the program adapter each time we create a compiler?

kyliau replied:
The adapter is stateful because it needs to keep track of the template versions.
We could make it stateless, but in that case, the language service class itself will have to keep track of the template versions. I think the responsibility of tracking versions should fall under the adapter, not the language service.

ayazhafiz replied:
Oh youre right I forgot about the versioning. Okay I am fine either way, sorry this was such a long discussion. Maybe later on we can figure out some nicer abstractions, even on the compiler side.

kyliau replied:
It's good, your comments make me think more about the architecture, and I agree there's an opportunity here to reuse some of TS logic to keep track of changed files, instead of us doing our custom checking. 
Could you please approve this PR?

ayazhafiz replied:
Of course :)

kyliau commented on packages/language-service/ivy/language_service_adapter.ts:
```diff
@@ -0,0 +1,112 @@
+    if (isTypeScriptFile(fileName)) {
+      throw new Error(`readResource() should not be called on TS file: ${fileName}`);
+    }
+    // Calling getScriptSnapshot() will actually create a ScriptInfo if it does
+    // not exist! The same applies for getScriptVersion().
+    // getScriptInfo() will not create one if it does not exist.
+    // In this case, we *want* a script info to be created so that we could
+    // keep track of its version.
+    const snapshot = this.project.getScriptSnapshot(fileName);
+    if (!snapshot) {
+      throw new Error(`Failed to get script snapshot while trying to read ${fileName}`);
+    }
```
yes, or if `readFile()` fails for whatever reasons. I've added a comment to clarify.
